### PR TITLE
Attempt to optimize updating the h3dex targeting indices

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 1]
```diff
@@ -4978,6 +4978,19 @@ lookup_gateways_from_hex(Hex, Ledger) when is_integer(Hex) ->
                          ]
               ).
 
+is_hex_populated(Hex, Ledger) ->
+    H3CF = h3dex_cf(Ledger),
+    cache_fold(Ledger, H3CF,
+               fun({_Key, _GWs}, _Acc) ->
+                       %% XXX I don't know if you can short circuit with a throw here
```
Comment (evanmcc) [inline src/ledger/v1/blockchain_ledger_v1.erl:4985]: you can't, you'll need to add to to the cf_fold behavior, and also make sure that the cleanup is happening at all the appropriate places.

**Comment by PaulVMo** [comment]
Looks good to me - I have it running on a follower and forced the random hex lookup to be built/maintained by changing the poc_targeting_verison checks in these functions to >=4. It is following the chain just fine and shows a ~25x savings by using the new add/remove hex to random lookup functions versus using the build_random_targeting_hex_lookup every time a new targeting hexes is added or removed.

```
2022-06-29 20:58:46.831 [info] <0.3977.0>@blockchain_ledger_v1:build_random_hex_targeting_lookup:{5160,13} build_random_hex_targeting_lookup in 7619 ms
2022-06-29 21:03:20.950 [info] <0.1958.0>@blockchain_ledger_v1:add_hex_to_random_lookup:{5122,12} add_hex_to_random_lookup in 219 ms
2022-06-29 21:03:38.490 [info] <0.1958.0>@blockchain_ledger_v1:add_hex_to_random_lookup:{5122,12} add_hex_to_random_lookup in 279 ms
2022-06-29 21:03:56.840 [info] <0.1958.0>@blockchain_ledger_v1:add_hex_to_random_lookup:{5122,12} add_hex_to_random_lookup in 252 ms
2022-06-29 21:04:17.332 [info] <0.1958.0>@blockchain_ledger_v1:add_hex_to_random_lookup:{5122,12} add_hex_to_random_lookup in 320 ms
2022-06-29 21:04:20.153 [info] <0.6980.0>@blockchain_ledger_v1:build_random_hex_targeting_lookup:{5160,13} build_random_hex_targeting_lookup in 6415 ms
```

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 2]
```diff
@@ -5021,6 +5044,83 @@ random_targeting_hex(RandState, Ledger) ->
             Error
     end.
 
+remove_hex_from_random_lookup(ParentRes, Ledger) ->
+    %% we only want to do this if poc version >= 4, which means h3dex targeting
```
Comment (andymck) [inline src/ledger/v1/blockchain_ledger_v1.erl:5048]:
```suggestion
    %% we only want to do this if poc version >= 6, which means h3dex targeting
```

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 3]
```diff
@@ -5021,6 +5044,83 @@ random_targeting_hex(RandState, Ledger) ->
...
+            {ok, <<Total:32/integer-unsigned-little>>} = cache_get(Ledger, H3CF, <<"population">>, []),
+           cache_put(Ledger, H3CF, <<"population">>, <<(Total-1):32/integer-unsigned-little>>),
+           ok;
+        _ ->
+            ok
+    end.
+
+
+add_hex_to_random_lookup(ParentRes, Ledger) ->
+    %% we only want to do this if poc version >= 4, which means h3dex targeting
```
Comment (andymck) [inline src/ledger/v1/blockchain_ledger_v1.erl:5086]:
```suggestion
    %% we only want to do this if poc version >= 6, which means h3dex targeting
```

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 4]
```diff
@@ -5021,6 +5044,83 @@ random_targeting_hex(RandState, Ledger) ->
...
+                              true;
+                          false ->
+                                HasMatched
+                      end;
+                 (_, Acc) ->
+                      Acc
+              end, false,
+              [
+               {start, {seek, <<"random-", 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0>>}},
+               {iterate_upper_bound, <<"random-ÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿ">>}
```
Comment (xandkar) [inline src/ledger/v1/blockchain_ledger_v1.erl:5113]: I see this (similar?) technique used in a few places - would be great to capture it as a reusable function with an attached explanatory comment on what the expected invariants are.

**Review by evanmcc — APPROVED** [review]

**Pull request merged**